noxfile.py

```diff
@@ -112,6 +112,7 @@ def build_docs(session: nox.Session) -> None:
     site_dir = session.posargs or ["site/"]
     session.install(*DOC_DEPENDENCIES, ".")
     session.env["PIPX__DOC_DEFAULT_PYTHON"] = "typically the python used to execute pipx"
+    session.run("towncrier", "build", "--version", "Next", "--keep")
```

Comment:
```suggestion
    session.run("towncrier", "build", "--draft", "--version", "Next", "--keep")
```

Why should we modify the changelog?

Reply:
If we don't,  the latest news fragments would not show up in the docs built, would they?

Reply:
Right, but the changelog is not the most important part of the documentation, so wouldn't it be enough to write to standard output? The risk of accidentally committing you mentioned would be gone with that.

Or we might add a `git restore` afterwards?

Reply:
@gaborbernat what's your opinion on this?

`git restore` seems dangerous to me in case someone actually does a fix in the changelog file. That would make them lose their changes.

Alternatively there's a mkdocs plugin to embed a markdown file in another. Using this, we could store the new changelog entries in a separate file with the towncrier build,  that is not checked into source control. Then we'd just embed it into the changelog. That's pretty much what `virtualenv` does in their sphinx configuration.

Reply:
Yeah the later is the path ahead.

Reply:
There's a couple of options for mkdocs plugins that allow this. I went for https://github.com/fralau/mkdocs-macros-plugin, because it has the most stars and comes with a lot of nice functionality that might be helpful in the future. It also had some of the easiest implementation for including the draft file.

Reply:
Where are the "live" docs served from? Are they also created with the build_docs nox command? If so, do I somehow need to check in the build step whether we're running on main or on a feature branch? Or do we want to add the upcoming news fragments to the `prod` changelog under a special section like `Upcoming` or `Planned for next Release`?

Reply:
> Where are the "live" docs served from? Are they also created with the build_docs nox command?

No, there's a separate session for that:

https://github.com/pypa/pipx/blob/4dd1b6b84cbc2dc14e958ff2223c7f3388982e98/noxfile.py#L119

> Or do we want to add the upcoming news fragments to the `prod` changelog under a special section like `Upcoming` or `Planned for next Release`?

Sounds reasonable.

Reply:
I've adapted it to show the upcoming changes separately at the top with a expressive header, both in dev and prod.

One thing to note is that the `_draft_changelog.md` file is copied into the sites directory. So while it's not linked anywhere, it is technically available if one enters it's URL directly. IMO that's fine. let me know if you disagree. Then I'll add deleting it from there to the nox command.

Reply:
We should delete it.

Reply:
> No, there's a separate session for that:
> 
> https://github.com/pypa/pipx/blob/4dd1b6b84cbc2dc14e958ff2223c7f3388982e98/noxfile.py#L119

@chrysle I'm quite confused by this. I can't see `watch_docs` be used anywhere in the repository. Can you give me a bit more detail on how it is used to serve our `prod` docs?

Reply:
> Can you give me a bit more detail on how it is used to serve our `prod` docs?

There isn't much to say about it. It builds the directory tree at it current state and serves it, rebuilding whenever a change is detected. So I think your modifications are right.

Reply:
> There isn't much to say about it. It builds the directory tree at it current state and serves it, rebuilding whenever a change is detected. So I think your modifications are right.

I assume we have a misunderstanding here, but correct me if I'm wrong.

- `build_docs` are used for serving the docs on `readthedocs`, including our production docs at https://pipx.pypa.io/
- `watch_docs` is for live previewing the docs locally during development.

Is that accurate?

If so - for `build_docs` I now remove the `_draft_changelog` from the site directory after building. For `watch_docs` that's not really easy to do because it builds and then serves immediately in a blocking way. No chance for me to delete the file in between. But if that is only used for dev - I think it does not matter. Might even be useful for debugging the behavior on dev.

Reply:
> Is that accurate?

Yes.

> But if that is only used for dev - I think it does not matter. Might even be useful for debugging the behavior on dev.

I also think so.